### PR TITLE
File info by name

### PR DESCRIPTION
**Pull request by przadka**
I included the file info by name method to the API. This is related to https://github.com/Backblaze/B2_Command_Line_Tool/issues/440 and I will follow up with another PR, updating the CLI tool.

I am not sure whether I am using the force_action parameter correctly:

```
return self.file_version_factory.from_api_response(
            self.session.get_file_info_by_name(bucket_name, file_name),
            force_action='folder',
        )
```

I did not manage to find what are allowed values for action so I picked something that seemed reasonable (as in: "I am listing the folder"). Please review my usage of force_action and let me know if there is a better way to handle this case

Probably the whole `def from_api_response(self, file_version_dict, force_action=None)` should have a better documentation and type hints.

### CODE REVIEW

File: b2sdk/file_version.py [hunk 1]
```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
...
-        else:
+
+        if force_action is not None:
+            file_version_dict['action'] = force_action
+
+        file_name = file_version_dict.get('fileName', file_version_dict.get('x-bz-file-name'))
+        id_ = file_version_dict.get('fileId', file_version_dict.get('x-bz-file-id'))
+
+        if file_name is None:
+            raise ValueError('no fileName or file')
```
Comment (agoncharov-reef) [inline b2sdk/file_version.py:518]: `no fileName or x-bz-file-name`

File: b2sdk/file_version.py [hunk 2]
```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
...
-            size = file_version_dict['contentLength']
-        else:
+
+        if force_action is not None:
+            file_version_dict['action'] = force_action
+
+        file_name = file_version_dict.get('fileName', file_version_dict.get('x-bz-file-name'))
+        id_ = file_version_dict.get('fileId', file_version_dict.get('x-bz-file-id'))
+
+        if file_name is None:
```
Comment (agoncharov-reef) [inline b2sdk/file_version.py:517]: I'd prefer an assertion here and below:
```python
assert file_name, "no fileName ..."
```

File: b2sdk/file_version.py [hunk 3]
```diff
@@ -525,8 +557,8 @@ def from_api_response(self, file_version_dict, force_action=None):
             content_sha1,
             file_info,
             upload_timestamp,
-            file_version_dict['accountId'],
-            file_version_dict['bucketId'],
+            file_version_dict.get('accountId'),
+            file_version_dict.get('bucketId'),
```
Comment (agoncharov-reef) [inline b2sdk/file_version.py:561]: Now it's less strict. Previous version required `accountId` to be in file_version_dict, current one requires nothing. It's because the method tries to handle two different sets of inputs.

File: b2sdk/file_version.py [hunk 4]
```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
...
-        if 'size' in file_version_dict:
-            size = file_version_dict['size']
-        elif 'contentLength' in file_version_dict:
-            size = file_version_dict['contentLength']
-        else:
+
+        if force_action is not None:
+            file_version_dict['action'] = force_action
+
+        file_name = file_version_dict.get('fileName', file_version_dict.get('x-bz-file-name'))
```
Comment (agoncharov-reef) [inline b2sdk/file_version.py:514]: Here and below: 
```python
file_name = file_version_dict.get('fileName') or file_version_dict.get('x-bz-file-name')
```

File: test/unit/api/test_api.py [hunk 5]
```diff
@@ -97,6 +97,37 @@ def test_get_file_info(self):
             assert isinstance(result, VFileVersion)
             assert result == created_file
 
+    def test_get_file_info_by_name(self):
```
Comment (agoncharov-reef) [inline test/unit/api/test_api.py:100]: There is no test for hidden file, for non-existent file, and for file with replication/SSE/legal hold.

File: b2sdk/api.py [hunk 6]
```diff
@@ -587,6 +587,18 @@ def get_file_info(self, file_id: str) -> FileVersion:
...
+    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> FileVersion:
+        """
+        Gets info about file version. Similar to `get-file-info` but 
+        takes the bucket name and file name instead of file id.
+
+        :param str file_id: the id of the file whose info will be retrieved.
+        """
+        return self.file_version_factory.from_api_response(
+            self.session.get_file_info_by_name(bucket_name, file_name),
+            force_action='folder',
```
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply b2sdk/api.py:599]: The guy in initial PR you mentioned says that `b2-download-file-by-name` will return 404 for hidden files, so most of the time `action` will be `upload`, but it could also be `start` as far as I understand. Probably there should be a test to start large file upload and try `HEAD b2-download-file-by-name`. If it returns 404 as well then I think it's safe to force `upload` action.
Dunno what is `folder` and how to deal with it.

[the remark this answers is not in the document]
Comment (przadka) [reply b2sdk/api.py:599]: ok, thanks! i actually missed the comment about hidden files in the original issue 🤦

File: b2sdk/file_version.py [hunk 7]
```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
...
-        file_name = file_version_dict['fileName']
-        id_ = file_version_dict['fileId']
-        if 'size' in file_version_dict:
-            size = file_version_dict['size']
-        elif 'contentLength' in file_version_dict:
-            size = file_version_dict['contentLength']
-        else:
+
+        if force_action is not None:
+            file_version_dict['action'] = force_action
```
[the remark this answers is not in the document]
Comment (przadka) [reply b2sdk/file_version.py:512]: it is used actually:

```
        file_retention = FileRetentionSetting.from_file_version_dict(file_version_dict)
        legal_hold = LegalHold.from_file_version_dict(file_version_dict)
```
both of these methods depend on action being properly set, for example:

```
 @classmethod
    def from_file_version_dict(cls, file_version_dict: dict) -> LegalHold:
        if 'legalHold' not in file_version_dict:
            if file_version_dict['action'] not in ACTIONS_WITHOUT_LOCK_SETTINGS:
                raise UnexpectedCloudBehaviour(
                    'legalHold not provided for file version with action=%s' %
                    (file_version_dict['action'])
                )
            return cls.UNSET
        if not file_version_dict['legalHold']['isClientAuthorizedToRead']:
            return cls.UNKNOWN
        return cls.from_string_or_none(file_version_dict['legalHold']['value'])
```
As I understand `if file_version_dict['action'] not in ACTIONS_WITHOUT_LOCK_SETTINGS` will give error if action is not set.
 
My understanding was that when we force the action, it is safe to assume that these methods should simply read from the dict, hence the medication of the input dictionary. Alternatively, I can pass force_action to these methods as a parameter. What would be your suggestion?

File: b2sdk/file_version.py (hunk 2 above)
Comment (przadka) [reply b2sdk/file_version.py:517]: Are we ok with changing the semantics from raising ValueError to AssertionError? Probably we are, I just want to confirm whether I should do some additional checking/reviewing before I change it.

File: b2sdk/file_version.py (hunk 3 above)
Comment (przadka) [reply b2sdk/file_version.py:561]: Indeed. I decided to change this because I noticed that file_version_dict can have different structure, depending on the exact path we take during the execution. Shall I assume you comment is just an observation or do you expect me to improve something?

File: b2sdk/file_version.py (hunk 4 above)
Comment (przadka) [reply b2sdk/file_version.py:514]: This was intentional because I experienced problem with short-circuiting of these boolean statements. I thought that using the get(...,get()) was more reliable and even more Pythonic way of doing this. 

To give you more context, when working with timestamps, I realized that when:
```
file_version_dict.get('uploadTimestamp'): 0
file_version_dict.get('x-bz-upload-timestamp'): None 
```
then `upload_timestamp = file_version_dict.get('uploadTimestamp' ) or file_version_dict.get('x-bz-upload-timestamp')` actually evaluates to `None` which I believe is wrong. This is why I changed this.

File: test/unit/api/test_api.py (hunk 5 above)
Comment (przadka) [reply test/unit/api/test_api.py:100]: ok, will write some more tests, once we are done with the basic review

File: b2sdk/file_version.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply b2sdk/file_version.py:512]: My bad, you're right. Still I think making fake `action` is not the way. I would rather modify `from_file_version_dict` for case when `action` is not there. Smth like:
```python
action = file_version_dict.get('action')
if action is not None and action not in ACTIONS_WITHOUT_LOCK_SETTINGS:
```

File: b2sdk/file_version.py (hunk 4 above)
Comment (agoncharov-reef) [reply b2sdk/file_version.py:514]: I didn't expect the opportunity for `uploadTimestamp` to be zero, but now I see that it's always 0 for action == `folder`. All good then

File: b2sdk/file_version.py [hunk 8]
```diff
@@ -488,23 +488,55 @@ def from_api_response(self, file_version_dict, force_action=None):
...
+
+            {
+                "content-length": 11, 
+                "content-type": "b2/x-auto", 
+                "x-bz-content-sha1": "2aae6c35c94fcfb415dbe95f408b9ce91ee846ed", 
+                "x-bz-upload-timestamp": 5000, 
+                "x-bz-file-id": "9999", 
+                "x-bz-file-name": "file"
+            }
+
```
[the remark this answers is not in the document]
Comment (przadka) [reply b2sdk/file_version.py:503]: @agoncharov-reef can you give me some more guidance on this? why do you think it should be two methods initially? ho would you call them? I am thinking about separating this into 3 methods, as per your suggestion, but as you can imagine `from_api_response` is a method that is called rather often :)

[the remark this answers is not in the document]
Comment (agoncharov-reef) [reply b2sdk/file_version.py:503]: I don't think you should untangle whatever was there before you, but maybe not tangle it even more - that's what I meant. And please correct me if I'm wrong below.

I see that `B2Session.get_file_info_by_name()` calls `B2RawHTTPApi` which generates download url and makes HEAD request to b2. So `B2Session.get_file_info_by_name` will always return headers from that download HEAD request, and nothing more:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> FileVersion:
        return self.file_version_factory.from_api_response(
            self.session.get_file_info_by_name(bucket_name, file_name),  <-- this is a response from HEAD
            force_action='folder',
        )
```

Compare it to `B2Session.get_file_info_by_id()` which under the hood makes a fair POST request and gets all information, not only headers:
```python
    def get_file_info(self, file_id: str) -> FileVersion:
        return self.file_version_factory.from_api_response(
            self.session.get_file_info_by_id(file_id)   <-- this is a response from POST
        )
```

So I thought that maybe if there are different types of data then you shouldn't process them with the same function. And my initial proposal was to create a `file_version_factory.from_api_head_response` method which would be able to create `FileVersion` just from headers information and nothing more, while `file_version_factory.from_api_response` would work with all the data from fair POST request.

---

However, during writing this I found that there is also `download_version_factory` in `B2Api` class:
```python
        self.file_version_factory = self.FILE_VERSION_FACTORY_CLASS(self)
        self.download_version_factory = self.DOWNLOAD_VERSION_FACTORY_CLASS(self)
```

I was wondering what it is, and here is its definition:
```python
class DownloadVersionFactory:
    """
    Construct :py:class:`b2sdk.v2.DownloadVersion` objects from download headers.
    """
```
But as I understand, "download headers" is exactly what we get when we call `B2Session.get_file_info_by_name()`. So if my assumption is correct, your `B2Api.get_file_info_by_name` could now look like:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> DownloadVersion:
        return self.download_version_factory.from_response_headers(
            self.session.get_file_info_by_name(bucket_name, file_name),
        )
```
However, now it returns `DownloadVersion` instead of file version. I wanted to solve this, so I searched in code and found `file_version_info_from_download_version` which converts `DownloadVersion` to `FileVersion` (and it also answers your question about `action`):
```python
def file_version_info_from_download_version(download_version: v2.DownloadVersion):
    return FileVersionInfo(
        id_=download_version.id_,
        file_name=download_version.file_name,
        size=download_version.size,
        content_type=download_version.content_type,
        content_sha1=download_version.content_sha1,
        file_info=download_version.file_info,
        upload_timestamp=download_version.upload_timestamp,
        action='upload',
        content_md5=None,
        server_side_encryption=download_version.server_side_encryption,
        file_retention=download_version.file_retention,
        legal_hold=download_version.legal_hold,
        cache_control=download_version.cache_control,
        api=download_version.api,
    )
```

So now your code could look something like this:
```python
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> FileVersion:
        download_version = self.download_version_factory.from_response_headers(
            self.session.get_file_info_by_name(bucket_name, file_name),
        )
        return file_version_info_from_download_version(download_version)
```

---

I was curious where `file_version_info_from_download_version` is used, and I found this in `v1/bucket.py::Bucket
```python
class Bucket:
    ...
    def get_file_info_by_name(self, file_name: str) -> FileVersionInfo:
        return file_version_info_from_download_version(super().get_file_info_by_name(file_name))
```

And that `super().get_file_info_by_name(...)` is this:
```python
    def get_file_info_by_name(self, file_name: str) -> DownloadVersion:
        """
        Gets a file's DownloadVersion by name.

        :param str file_name: the name of the file who's info will be retrieved.
        """
        try:
            return self.api.download_version_factory.from_response_headers(
                self.api.session.get_file_info_by_name(self.name, file_name)
            )
        except FileOrBucketNotFound:
            raise FileNotPresent(bucket_name=self.name, file_id_or_name=file_name)
```

And all this is exactly what I proposed: get headers from `get_file_info_by_name`, convert them to `DownloadVersion`, then convert it to `FileVersion`.

So it looks like `v1.Bucket.get_file_info_by_name()` returns `FileVersionInfo`, while latest `Bucket.get_file_info_by_name()` returns `DownloadVersion`. I think the latter is more correct, since you cannot recreate full file version from headers; and it's newer anyway. So let's use it!

```python
class B2Api:
    ...
    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> DownloadVersion:
        bucket = self.get_bucket_by_name(bucket_name)
        return bucket.get_file_info_by_name(file_name)
```

File: test/unit/api/test_api.py (hunk 5 above)
Comment (przadka) [reply test/unit/api/test_api.py:101]: @agoncharov-reef added new tests

File: b2sdk/api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (przadka) [reply b2sdk/api.py]: @agoncharov-reef looks like `Bucket.get_file_info_by_name()` works correctly for hidden files as well

File: b2sdk/file_version.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (przadka) [reply b2sdk/file_version.py]: thanks a lot for this comprehensive review. I reverted back to the previous version 'file_info.py` and now I am using `Bucket.get_file_info_by_name()` to pull the information. there are much less changes in the code currently

File: test/unit/api/test_api.py [hunk 9]
```diff
@@ -100,15 +101,15 @@ def test_get_file_info(self):
...
             b'hello world',
             'file',
         )
         result = self.api.get_file_info_by_name('bucket1', 'file')
 
         expected_result = {
-            'fileId': '9999',
-            'fileName': 'file',
+            'fileId': new_file.id_,
+            'fileName': new_file.file_name,
```
Comment (agoncharov-reef) [inline test/unit/api/test_api.py:112]: I think that hardcoding values (where possible) in tests is better than using something dynamic, since that "something" may be wrong. I'd prefer previous version :)

File: test/unit/api/test_api.py [hunk 10]
```diff
@@ -135,6 +136,74 @@ def test_get_file_info_by_name(self):
...
+            'contentSha1': 'none',
+            'replicationStatus': None,
+        }
+
+        if apiver_deps.V <= 1:
+            expected_result.update({
+                'accountId': None,
+                'action': 'upload',
+                'bucketId': None,
+            })
```
Comment (agoncharov-reef) [inline test/unit/api/test_api.py:173]: Nice!

File: b2sdk/api.py [hunk 11]
```diff
@@ -587,6 +593,16 @@ def get_file_info(self, file_id: str) -> FileVersion:
             self.session.get_file_info_by_id(file_id)
         )
 
+    def get_file_info_by_name(self, bucket_name: str, file_name: str) -> DownloadVersion:
+        """
+        Gets info about file version. Similar to `get_file_info` but 
+        takes the bucket name and file name instead of file id.
+
+        :param str file_id: the id of the file whose info will be retrieved.
```
Comment (agoncharov-reef) [inline b2sdk/api.py:601]: This is misleading

File: b2sdk/file_lock.py [hunk 12]
```diff
@@ -108,7 +108,9 @@ def from_file_version_dict(cls, file_version_dict: dict) -> FileRetentionSetting
             }
         """
         if 'fileRetention' not in file_version_dict:
-            if file_version_dict['action'] not in ACTIONS_WITHOUT_LOCK_SETTINGS:
+            action = file_version_dict.get('action')
+            if action is not None and file_version_dict['action'
+                                                       ] not in ACTIONS_WITHOUT_LOCK_SETTINGS:
```
Comment (agoncharov-reef) [inline b2sdk/file_lock.py:113]:
```python
if action is not None and action not in ACTIONS_WITHOUT_LOCK_SETTINGS
```
but is this change required at all?

File: b2sdk/api.py (hunk 11 above)
Comment (przadka) [reply b2sdk/api.py]: :facepalm:

**Comment by mjurbanski-reef** [comment]
already squash merged  to upstream
